AP_MotorsMulticopter: support motor spool down

rdehaggart1 opened the pull request:
As discussed in [this forum discussion](https://discuss.ardupilot.org/t/mot-spool-time-not-working-as-expected-when-drone-is-disarmed-v4-5-7/130947), the 'spooling' function was not behaving as expected when disarming.

### Symptoms:
When setting _MOT_SPOOL_TIME_ (and leaving _MOT_SPOOL_TIM_DN=0_), the ramp up was handled well, but there was no ramping down (see below)

<img width="854" height="526" alt="beforeFix" src="https://github.com/user-attachments/assets/ffbc0546-b417-4216-9c4f-bf90fbc84ebc" />

### Fix:
A single line of code was forcing ``_spool_state = SpoolState::SHUT_DOWN`` when disarmed. This meant that instead of progressing through the normal state transitions, the motors were being forced straight into stopping without any ramp logic ever executing.
By removing this line, but keeping the line ``_spool_desired = DesiredSpoolState::SHUT_DOWN;``, the state controller still targets SHUT_DOWN but transitions properly into it.

<img width="854" height="526" alt="afterFix" src="https://github.com/user-attachments/assets/5afa5440-dbd2-423a-a899-272d8b8a9ec3" />

### Additional changes
As part of this PR, I also reviewed and updated some of the comments associated with the spooling logic in ``AP_MotorsMulticopter::output_logic()``. This was done to update incorrect comments, and to improve the clarity of the code in this function.

rdehaggart1 commented:
In an admittedly quite specific use case (with an ICE and high inertia rotors), the 'step' off behaviour caused a torque reduction that was too sudden. 
In trying to solve this problem, I noticed the 'spool' parameters. They describe themselves in a way that sounds as if the intended behaviour is as shown in the second data plot above. 

The reason I thought this was a bug, is that the MOT_SPOOL_TIM_DN parameter was having no effect, and the code that is implemented between GROUND_IDLE and SHUT_DOWN was never being executed. 

I understand what you're saying about the arm behaviour, however.

peterbarker commented:
> In an admittedly quite specific use case (with an ICE and high inertia rotors), the 'step' off behaviour caused a torque reduction that was too sudden. In trying to solve this problem, I noticed the 'spool' parameters. They describe themselves in a way that sounds as if the intended behaviour is as shown in the second data plot above.

They're definitely there exactly for the reason you're suggesting!

> I understand what you're saying about the arm behaviour, however.

Could you check if using an interlock switch fixes your problem, please?

rdehaggart1 commented:
## Update:

I have made a couple of modifications and commited to this PR

### **Mod No.1**: Separated the behaviours of 'disarm' and 'interlock disable' in the spooling state machine. 

This now forces a transition straight to SHUT_DOWN on disarm command, but allows the spool down logic to execute when commanding a motor interlock disable. 

### **Mod No.2**: Modified Arducopter/Radio.cpp to change the conditions for set_throttle_zero_flag().

This function sets a flag for when throttle goes non-zero. Once this flag is set, the spool state machine advances to THROTTLE_UNLIMITED. As far as I can tell, this makes sense for trad heli operations due to the slightly different states and startup procedures, but it means that a multicopter wouldn't ever have the motors spinning at MOT_SPIN_ARM (i.e. in GROUND_IDLE state) if using an interlock. 

For the purposes that this PR is trying to achieve, I added an extra condition such that if the frame is a multicopter, then the flag is only set once the throttle input is greater than zero. The conditions should be unaffected for trad heli and the operation should not have changed. (I say should because I dont have a heli to test on). 

### Results

Displayed in the below graph is the new behaviour on this build. The motors spool up and down upon the toggling of the interlock. As shown in the previous comment in this PR, disarming still forces a 'step' off behaviour.
<img width="854" height="526" alt="finalBehaviour" src="https://github.com/user-attachments/assets/16efebd2-bdfc-4b1f-b9df-ba64539562a1" />

rdehaggart1 commented:
> We discussed this at DevCall this-morning:

Thanks Peter, I appreciate the minutes!

> I was the heli maintainer until ~2017 and wrote most of the spool up/down, E-stop and Interlock as it existed then.

Thanks very much for taking the time to provide your input!

It's interesting to hear your description of the spooling function, and it sounds more or less as I would expect in terms of the intended functionality. 

> First and foremost, I want to say that IMO, under no circumstances should Ardupilot ever be directing any motor controllers or other energy output devices to output energy, when the state is Disarmed. This is a core safety concept.

Totally agree with this sentiment. When I published the first PR it completely slipped my mind that **even though** the PWM command was decreasing, that it was still being actively driven, and that is a safety issue. My apologies for not recognising that sooner - as I continue to learn the codebase and make contributions it will be at the forefront of my mind!

> Whatever happens, I'd like to not break the way this works for helis, because I think the old method of operation that I described above is still relevant for many hobbyists trying to apply Ardupilot to a standard hobby helicopter.

I also completely understand this, and I believe my changes have not impacted the behaviour for heli.

> When we brought the Motor Interlock and E-stop functions to multirotors, maybe that ramp-down period wasn't used, because shutting off the motors immediately wasn't an issue for multirotors in use at that time, and maybe this needs to be revisited.

I think this is the root. The current logic in master for multirotors (AP_Motors/MotorsMulticopter.cpp) is:

```
// force desired and current spool mode if disarmed or not interlocked
if (!armed() || !get_interlock()) {
    _spool_desired = DesiredSpoolState::SHUT_DOWN;
    _spool_state = SpoolState::SHUT_DOWN;
}
```

This means that for multicopters, when disabling interlock, the spool down from GROUND_IDLE or from THROTTLE_UNLIMITED is bypassed. 

In AP_Motors/MotorsHeli.cpp:
```    
// force desired and current spool mode if disarmed and armed with interlock enabled
if (armed()) {
    if (!get_interlock()) {
        _spool_desired = DesiredSpoolState::GROUND_IDLE;
    } else {
        ..
    }
} else {
    ..
    _spool_desired = DesiredSpoolState::SHUT_DOWN;
    _spool_state = SpoolState::SHUT_DOWN;
}
```

So the case of disarming is forced to SHUT_DOWN, but the case of disabling interlock goes via the spooling state machine. 

---

### Summary
Looking back at the current state of the PR:
1. I have reverted my changes such that disarming still forces the state machine to SHUT_DOWN, bypassing the spool logic. This retains the safety rule and moves to disable the motor drive immediately on disarm
2. I have separated the interlock behaviour from the arm behaviour in MotorsMultirotor.cpp such that disabling interlock can transition through the state machine via the spooling logic.
3. I have modified Arducopter/radio.cpp to prevent multirotors from transitioning to THROTTLE_UNLIMITED when still on the ground and with zero throttle input. I believe I have kept this behaviour separated from heli such that the heli behaviour should remain as it was, though I'm happy to be challenged on that if there are concerns and we can work through it.

For anyone that wasn't aware, the origin of this arose from quite a specific multicopter use case. We were working with a very large platform that used an ICE and very large rotors with lots of intertia. Even with the rotors spinning at MOT_SPIN_ARM, the sudden step off to no load was too fast for the ICE. The spool down function would be much kinder to the transmission in this use case. I do also understand if this use case is too specific for the broader code to be changed to suit it.

Thanks to everyone for the detailed and engaging input on this!

lthall reviewed (commented):
I am not happy with these changes going in. There are too many behavioural changes here. I am also not comfortable with so many comment changes making it hard to review the behavioural changes.

Changes in this part of the code need a lot more planning and consultation before committing to a PR. The PR is not the place to discuss what the PR should be, that should be done in the issue or feature request.

lthall commented:
I have been working through the code I am responsible for in an attempt to improve the comments so I am always very happy when an area of the code requiring that work is highlighted.

Given the questions raised in this PR there is obviously a lot missing from the comments. I have tried to do a more through job here: #31380. 

As I said in the dev call I am happy to get on a call with you to discuss the requirements and restrictions we need to work around.
